### PR TITLE
Introduce a real logging library

### PR DESCRIPTION
**Pull request by joelsmithTT**
### Issue
Follow on from https://github.com/tenstorrent/tt-umd/pull/176
https://github.com/tenstorrent/tt-umd/issues/160
https://github.com/tenstorrent/tt-umd/issues/140

### Description
* Adds a real logging library [spdlog](https://github.com/gabime/spdlog) to replace the homegrown logger.hpp
* Lazy initializes
* Supports logging to console (stderr) and/or disk
* Not integrated yet (does not yet replace logger.hpp) -- want feedback from the team prior to replacing existing log statements

### List of the changes
* Introduces new logging API
* Adds tests and performance characterization code 

### Testing
* Unit tests

### API Changes
There are no public API changes in this PR.

### CODE REVIEW

File: cmake/dependencies.cmake [hunk 1]
```diff
@@ -94,22 +94,27 @@ function(fetch_dependencies)
...
-    ############################################################################################################################
+    ###################################################################################################################
     if(MASTER_PROJECT)
         CPMAddPackage(NAME nanobench GITHUB_REPOSITORY martinus/nanobench GIT_TAG v4.3.11)
     endif()
+
+    ####################################################################################################################
+    # spdlog
+    ####################################################################################################################
+    CPMAddPackage(NAME spdlog GITHUB_REPOSITORY gabime/spdlog GIT_TAG v1.14.1 VERSION v1.14.1)
```
Comment (broskoTT) [inline cmake/dependencies.cmake:118]: Is this used somewhere else in our repos, or are we the first to introduce the usage?

File: tests/misc/test_logger.cpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        buf << file.rdbuf();
+        return buf.str();
+    }
+};
+
+TEST_F(LoggerTest, BasicLogging) {
+    // Initialize logger with our test configuration
+    tt::umd::logger::Options options;
+    options.log_to_stderr = false;
+    options.filename = log_file.string();
```
Comment (broskoTT) [inline tests/misc/test_logger.cpp:54]: Does this still print out to stdout?

File: device/CMakeLists.txt [hunk 3]
```diff
@@ -21,6 +21,7 @@ set(UMD_DEVICE_SRCS
     wormhole/wormhole_coordinate_manager.cpp
     pcie/pci_device.cpp
     hugepage.cpp
+    ../common/logger_.cpp # ../common looks a bit strange
```
Comment (broskoTT) [inline device/CMakeLists.txt:24]: Do we intend to remove the other logger.hpp ?
And to rewrite all logs in tt_silicon_driver?

Comment (joelsmithTT) [reply device/CMakeLists.txt:24]: Yes/yes, just not in this PR.
Wanted to get the basic infrastructure reviewed by the team before imposing it everywhere.

File: cmake/dependencies.cmake (hunk 1 above)
Comment (joelsmithTT) [reply cmake/dependencies.cmake:118]: We're the first as far as I know

**Review by pjanevskiTT — COMMENTED** [review]
Nice changes in general. I don't have strong opinions on what logger we are using, I just have 3 things that I would like to see in order to consider this as an improvement or at least not a regression

- Consistency across repos
- Logging performance
- Build performance

I have left separate comments for all of this in the review, so we can discuss it in a thread. Let me know how those sound to you

File: cmake/dependencies.cmake (hunk 1 above)
[the remark this answers is not in the document]
Comment (joelsmithTT) [reply cmake/dependencies.cmake:118]: I think it's fmt that slows down the build.  spdlog uses fmt (and so do we already).
`main` build:
```
[304/304] Linking CXX executable test/umd/galaxy/unit_tests_glx

real	0m20.335s
user	5m3.613s
sys	0m14.538s
```
`joelsmith/spdlog-v2` build (which includes old logger)
```
[314/314] Linking CXX executable test/umd/galaxy/unit_tests_glx

real	0m20.353s
user	5m37.568s
sys	0m15.509s
```
... but I have a big system ;)

File: tests/misc/test_logger.cpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+using namespace tt::umd::util;
+
+class LoggerTest : public ::testing::Test {
+protected:
+    std::filesystem::path temp_dir;
+    std::filesystem::path log_file;
+
+    void SetUp() override {
+        // A bit of a hack:
```
[the remark this answers is not in the document]
Comment (joelsmithTT) [reply tests/misc/test_logger.cpp:23]: In typical usage the logger won't ever be disabled/re-enabled in an application's lifetime, but in the unit test (which runs a single process for multiple separate tests) the flag needs to be reset.

File: common/logger_.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * Macros for using the logger.
+ */
+#define UMD_TRACE(...)                                   \
+    do {                                                 \
+        ::tt::umd::logger::detail::ensure_initialized(); \
+        SPDLOG_TRACE(__VA_ARGS__);                       \
+    } while (0)
+
+#define UMD_DEBUG(...)                                   \
```
[the remark this answers is not in the document]
Comment (joelsmithTT) [reply common/logger_.hpp:46]: Is the concern just that you want the statements to be identical?  If so, I think we could aspire to a tt-common repo in which we accumulate software capital: tested, reusable components that any project can use.  (I have some reservations about doing this right now - my immediate want is to have source locations in our logging for my own development...)

There's a bunch of inconsistencies in our cross-project logging, and Metal/UMD have separate (but ancestrally related) implementations.
* Metal's logger is not thread safe (!); UMD's is
* Metal has its own env variables controlling logger behavior
* Metal supports e.g. `log_debug(type, fmtmsg)` _and_ `log_debug(fmtmsg)` consistently between levels, UMD is a bit of a mess here (`log_debug` needs a type, `log_info` does not)

FWIW I am not attached to `UMD_DEBUG` `UMD_INFO` etc. -- they could certainly be TT_ prefixed instead.

File: tests/misc/test_logger.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
+/*
```
[the remark this answers is not in the document]
Comment (joelsmithTT) [reply tests/misc/test_logger.cpp:1]: Thanks.  I had some help from an LLM to generate the boilerplate - it didn't get everything right but it sure saved some effort on my part! :)

**Comment by blozano-tt** [comment]
> One more thing I forgot to mention that would be extremely nice to have is env variable to control the level of log that is supported. Basically just control the `Option` for logging level

IE something that a logging library would give us for free?

https://github.com/gabime/spdlog/blob/v1.x/include/spdlog/cfg/env.h

File: tests/misc/test_logger.cpp [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    auto log_content = read_log_file();
+
+    EXPECT_TRUE(log_content.find("First message") != std::string::npos);
+    EXPECT_TRUE(log_content.find("Second message") != std::string::npos);
+    EXPECT_EQ(log_content.find("DIFFERENT:"), std::string::npos);
+}
+
+/**
+ * The next few tests aren't really unit tests - just a mechanism to understand
+ * the performance of the logger.  A log message that isn't printed (i.e. the
```
[the remark this answers is not in the document]
Comment (joelsmithTT) [reply tests/misc/test_logger.cpp:140]: I'm reluctant to spend time doing this if we're going to replace it.

[the remark this answers is not in the document]
Comment (pjanevskiTT) [reply tests/misc/test_logger.cpp:140]: makes sense

File: tests/misc/test_logger.cpp (hunk 2 above)
Comment (broskoTT) [reply tests/misc/test_logger.cpp:54]: Maybe add a comment that this is printing to a file only. Or add printing to both.

File: common/logger_.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (broskoTT) [reply common/logger_.hpp:46]: related thread: https://tenstorrent.slack.com/archives/C03N2GED3PS/p1731519222630709

File: tests/misc/test_logger.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (joelsmithTT) [reply tests/misc/test_logger.cpp:23]: I added a comment

**Comment by joelsmithTT** [comment]
For cross-project consistency, there's a design decision around log identification.
current behavior in this implementation is:
[2024-11-14 11:00:48.045] [info] [test_logger.cpp:184] Test message

If this becomes a shared solution across projects, we'd want to distinguish the source:
[2024-11-14 11:00:48.045] [UMD] [info] [test_logger.cpp:184] Test message
[2024-11-14 11:00:48.045] [METAL] [info] [somewhere_else.cpp:123] Other message

This would affect macro naming/use and initialization.

I suggest we focus on establishing robust logging into UMD first -- this will directly improve UMD developer experience by:
* eliminating [conditional compilation ](https://github.com/tenstorrent/tt-umd/blob/e03bc51e2aa03feb24f0f2f06697f298f06fda23/common/logger.hpp#L244C8-L244C24) affecting UMD logger behavior
* providing compile-time checks for string formatting in log messages
* showing source location in log messages

Later on, after we've lived with the spdlog for a while and have had opportunity to adjust how we are using it -- _then_ consider how to achieve consistency across projects.  Thoughts?

**Review by broskoTT — APPROVED** [review]
Seems like we'll feel benefits by using external well established logging library.
As long as this is not going in some opposite direction (which doesn't seem like it is the case) to other repos, I'm all for it.

File: common/logger_.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (broskoTT) [reply common/logger_.hpp:46]: > What I liked about our previous logger was that the logging was the same in every project

Hopefully we'll end up at the same place...

[the remark this answers is not in the document]
Comment (joelsmithTT) [reply common/logger_.hpp:46]: I'm going to merge this as-is so I can start using it.  I've captured some future work items around logging in the following issues.  Please take a look and add your input if you get an opportunity.

https://github.com/tenstorrent/tt-umd/issues/315 <- going to defer settling on the names of the macros until this
https://github.com/tenstorrent/tt-umd/issues/316

**Comment by blozano-tt** [comment]
Yayy! Thanks @joelsmithTT !

**Pull request merged**